examples/pennylane_run_rotoselect.py

```diff
@@ -0,0 +1,431 @@
+# for alternative gate choices.
+
+from matplotlib import cm
+from matplotlib.ticker import MaxNLocator
+from mpl_toolkits.mplot3d import Axes3D
+
+fig = plt.figure(figsize=(6, 4))
+ax = fig.gca(projection="3d")
+
+X = np.linspace(-4.0, 4.0, 40)
+Y = np.linspace(-4.0, 4.0, 40)
+xx, yy = np.meshgrid(X, Y)
+Z = np.array([[cost([x, y]) for x in X] for y in Y]).reshape(len(Y), len(X))
+surf = ax.plot_surface(xx, yy, Z, cmap=cm.coolwarm, antialiased=False)
+
+ax.set_xlabel(r"$\theta_1$")
+ax.set_ylabel(r"$\theta_2$")
+ax.zaxis.set_major_locator(MaxNLocator(nbins=5, prune="lower"))
+
+plt.show()
+
+##############################################################################
+# Rotoselect
+# ----------
+#
+# .. figure:: ../../examples/figures/rotoselect_structure.png
+#    :scale: 65%
+#    :align: center
+#    :alt: rotoselect_structure
+#
+# |
+# We now implement the Rotoselect algorithm to learn a good selection of gates to minimize
+# our cost function. The structure is similar to the original ansatz, but the generators of rotation are
+# selected from the set of Pauli gates :math:`P_d \in \{X,Y,Z\}` as shown in the figure above. For example,
+# :math:`U(\theta,Z) = R_z(\theta)`.
+#
+# Creating a quantum circuit with variable gates
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# First, we set up a quantum circuit with a similar structure to the one above, but
+# instead of fixed rotation gates :math:`R_x` and :math:`R_y`, we allow the gates to be specified with the
+# ``generators`` keyword. A helper method ``RGen`` returns the correct unitary gate according to the
+# rotation specified by an element of ``generators``.
+
+
+def RGen(param, generator, wires):
+    if generator == "X":
+        qml.RX(param, wires=wires)
+    elif generator == "Y":
+        qml.RY(param, wires=wires)
+    elif generator == "Z":
+        qml.RZ(param, wires=wires)
+
+
+def ansatz(params, generators):
+    RGen(params[0], generators[0], wires=0)
+    RGen(params[1], generators[1], wires=1)
+    qml.CNOT(wires=[0, 1])
+
+
+@qml.qnode(dev)
+def circuit(params, generators=[]):  # generators will be passed as a keyword arg
+    ansatz(params, generators)
+    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))
+
+
+@qml.qnode(dev)
+def circuit2(params, generators=[]):  # generators will be passed as a keyword arg
+    ansatz(params, generators)
+    return qml.expval(qml.PauliX(0))
+
+
+def cost(params, generators):
+    Z_1, Y_2 = circuit(params, generators=generators)
+    X_1 = circuit2(params, generators=generators)
+    return 0.5 * Y_2 + 0.8 * Z_1 - 0.2 * X_1
+
+
+##############################################################################
+# Helper methods
+# ~~~~~~~~~~~~~~
+# We define helper methods in a similar fashion to Rotosolve. In this case,
+# we must iterate through the possible gate choices in addition to optimizing each parameter.
+
+
+def rotosolve(d, params, generators, cost, M_0):  # M_0 only calculated once
```

Comment:
are you passing the cost function in the arguments?

Reply:
Yes, consistent with rotosolve above and PennyLane optimizers in general right?

Reply:
yeah but Rotoselcet and rotosolve do a simple for loop over the required pre-defined functions. I don't see the use of passing the cost function and fear it may be confusing since so many variables are called cost. I am going to leave the final decision up to @cgogolin :)

Reply:
Sounds good. One final thing to note is that the same can be said of the `params` list as well; there's no real need to pass it into the function as it is mutable. However, the behaviour of `GradientDescent` and other optimizers is that the `step` method takes `params` and `cost` as arguments and returns the updated `params`, so I was working in analogy with this.